### PR TITLE
Tool modifications to verify premium packages and metadata.

### PR DESCRIPTION
**Pull request by noya3110**
Code modifications to verify premium packages and metadata.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -137,5 +139,22 @@ Certificates verified successfully!
 Metadata Signature Verified successfully!
 ```
 
+#### Example 2
```
Comment (satyam-aoss) [inline README.md:142]: should we rename from Example 2(heading) to Example to verify premium packages? and similarly rename example 1

File: cmd/set_config.go [hunk 2]
```diff
@@ -15,79 +15,75 @@
 package cmd
 
 import (
-    "fmt"
-    "log"
-    "os/user"
-    "path/filepath"
```
Comment (satyam-aoss) [inline cmd/set_config.go:21]: looks like this file has changes in tab space, please check this once.

File: cmd/utils.go [hunk 3]
```diff
@@ -15,378 +15,411 @@
...
+		return fmt.Errorf("failed to create file: %v", err)
+	}
+	defer file.Close()
+
+	if _, err := io.Copy(file, reader); err != nil {
+		return fmt.Errorf("failed to download the file: %v", err)
+	}
+
+	// Close the client.
+	client.Close()
```
Comment (satyam-aoss) [inline cmd/utils.go:133]: can we move this client.Close to line 115 and also add defer?

File: README.md (hunk 1 above)
Comment (noya3110) [reply README.md:142]: done.

File: cmd/utils.go (hunk 3 above)
Comment (noya3110) [reply cmd/utils.go:133]: done

File: cmd/set_config.go (hunk 2 above)
Comment (noya3110) [reply cmd/set_config.go:21]: done

File: cmd/utils.go [hunk 4]
```diff
@@ -180,14 +213,62 @@ func parseBuildInfoJSON(jsonFile, spdxID string) (sigURL, gcpKmsKey string, buil
...
+            }
+        }
+    }
+
+    // Get build provenance signatures and public key.
+    if len(jsonData.BuildDetails) < 1 {
+        return SigDetails{}, "", nil, fmt.Errorf("couldn't get build details")
+    }
+    if len(jsonData.BuildDetails[0].BuildProvenances) < 1 {
+        return SigDetails{}, "", nil, fmt.Errorf("couldn't get build provenance details")
```
Comment (satyam-aoss) [inline cmd/utils.go:260]: we can make sigdetails as pointer, and send it as nil

File: cmd/verify_metadata.go [hunk 5]
```diff
@@ -87,28 +86,35 @@ func verifyMetadata(cmd *cobra.Command, args []string) error {
...
     if err != nil {
         return err
     }
+
+    metadata_type, err := cmd.Flags().GetString(metadataTypeFlagName)
+    if err != nil {
+        return err
+    }
+
+    if metadata_type != "buildinfo" && metadata_type != "vexinfo" && metadata_type != "healthinfo" && metadata_type != "premiuminfo" {
```
Comment (satyam-aoss) [inline cmd/verify_metadata.go:109]: we can declare an array, and then check if metadata_type is present in that array or not.

File: cmd/verify_metadata.go [hunk 6]
```diff
@@ -144,41 +150,137 @@ func verifyMetadata(cmd *cobra.Command, args []string) error {
...
     }
 
+    if metadata_type == "premiuminfo" {
+        return verifyPremiumMetadata(cmd, serviceAccountKeyFilePath, destDir, artifactPath, language, packageID, version, disableCertificateVerification, disableDeletes)
+    }
+
+    return verifyNONPremiumMetadata(cmd, serviceAccountKeyFilePath, destDir, language, packageID, version, metadata_type, disableCertificateVerification, disableDeletes)
+}
+
+func verifyPremiumMetadata(cmd *cobra.Command, serviceAccountKeyFilePath, destDir, artifactPath, language, packageID, version string, disableCertificateVerification, disableDeletes bool) error {
```
Comment (satyam-aoss) [inline cmd/verify_metadata.go:160]: there are lot of inputs fields, we can make a struct and take it as an input.

Comment (satyam-aoss) [reply cmd/verify_metadata.go:160]: you can also add a comment, on how you are verifying the premium metadata.

File: cmd/verify_metadata.go [hunk 7]
```diff
@@ -144,41 +150,137 @@ func verifyMetadata(cmd *cobra.Command, args []string) error {
...
+            if err != nil {
+                return err
+            }
+        }
+
+    }
+    return nil
+}
+
+func verifyNONPremiumMetadata(cmd *cobra.Command, serviceAccountKeyFilePath, destDir, language, packageID, version, metadata_type string, disableCertificateVerification, disableDeletes bool) error {
```
Comment (satyam-aoss) [inline cmd/verify_metadata.go:241]: we can create struct for input. Also can you check if verifyNONPremiumMetadata and verifyPremiumMetadata shares any common code that can be abstract as a separate function.

File: cmd/verify_package.go [hunk 8]
```diff
@@ -162,40 +158,174 @@ func verifyPackage(cmd *cobra.Command, args []string) error {
...
     destDir := fmt.Sprintf("%s-%s-%s", packageID, version, time.Now().Format("2006_01_02_15:04:05"))
     destDir = filepath.Join(downloadsDir, destDir)
     if err := os.Mkdir(destDir, os.ModePerm); err != nil {
         return err
     }
 
+    return verifyPremiumPackage(cmd, destDir, serviceAccountKeyFilePath, artifactPath, language, packageID, version, disableDeletes, verifyBuildProvenance, disableCertificateVerification)
+}
+
+func verifyPremiumPackage(cmd *cobra.Command, destDir, serviceAccountKeyFilePath, artifactPath, language, packageID, version string, disableDeletes, verifyBuildProvenance, disableCertificateVerification bool) error {
```
Comment (satyam-aoss) [inline cmd/verify_package.go:171]: inputs can be in a single struct.

File: cmd/verify_package.go [hunk 9]
```diff
@@ -162,40 +158,174 @@ func verifyPackage(cmd *cobra.Command, args []string) error {
...
+    if !disableDeletes {
+        destDir = strings.TrimSuffix(destDir, "/metadata.json")
+        if err := os.RemoveAll(destDir); err != nil {
+            return err
+        }
+    }
+    return nil
+}
+
+func verifyNONPremiumPackage(cmd *cobra.Command, destDir, serviceAccountKeyFilePath, artifactPath, language, packageID, version string, disableDeletes, verifyBuildProvenance, disableCertificateVerification bool) error {
```
Comment (satyam-aoss) [inline cmd/verify_package.go:295]: inputs can be in a single struct. And for package as well, check if something from this function and the previous function can be abstracted out.

**Review by satyam-aoss — APPROVED** [review]

File: cmd/verify_metadata.go [hunk 10]
```diff
@@ -144,41 +150,137 @@ func verifyMetadata(cmd *cobra.Command, args []string) error {
         return err
     }
 
+    if metadata_type == "premiuminfo" {
+        return verifyPremiumMetadata(cmd, serviceAccountKeyFilePath, destDir, artifactPath, language, packageID, version, disableCertificateVerification, disableDeletes)
+    }
+
+    return verifyNONPremiumMetadata(cmd, serviceAccountKeyFilePath, destDir, language, packageID, version, metadata_type, disableCertificateVerification, disableDeletes)
```
Comment (satyam-aoss) [inline cmd/verify_metadata.go:157]: intead of non premium, should be re-name it to standard

**Review by satyam-aoss — APPROVED** [review]

File: cmd/verify_metadata.go (hunk 10 above)
Comment (noya3110) [reply cmd/verify_metadata.go:157]: done

File: cmd/verify_package.go (hunk 8 above)
Comment (noya3110) [reply cmd/verify_package.go:171]: done

File: cmd/verify_package.go (hunk 9 above)
Comment (noya3110) [reply cmd/verify_package.go:295]: done

File: cmd/verify_metadata.go (hunk 6 above)
Comment (noya3110) [reply cmd/verify_metadata.go:160]: done

File: cmd/verify_metadata.go (hunk 7 above)
Comment (noya3110) [reply cmd/verify_metadata.go:241]: done.

File: cmd/verify_metadata.go (hunk 5 above)
Comment (noya3110) [reply cmd/verify_metadata.go:109]: done

File: cmd/utils.go (hunk 4 above)
Comment (noya3110) [reply cmd/utils.go:260]: done

**Pull request merged**